Realm/Realm/Native/SyncSocketProvider.EventLoop.cs

```diff
@@ -0,0 +1,154 @@
+            private readonly Status _status;
+
+            // Belongs to SyncSocketProvider. When Native destroys the Provider we need to stop executing
+            // enqueued work, but we need to release all the callbacks we copied on the heap.
+            private readonly CancellationToken _cancellationToken;
+
+            public EventLoopWork(IntPtr nativeCallback, Status status, CancellationToken cancellationToken)
+            {
+                _nativeCallback = nativeCallback;
+                _status = status;
+                _cancellationToken = cancellationToken;
+            }
+
+            public unsafe void Execute()
+            {
+                if (_cancellationToken.IsCancellationRequested)
+                {
+                    Logger.LogDefault(LogLevel.Trace, "Deleting EventLoopWork callback only because event loop was cancelled.");
+                    NativeMethods.delete_callback(_nativeCallback);
+                    return;
+                }
+
+                RunCallback(_nativeCallback, _status);
+            }
+        }
+
+        private static void RunCallback(IntPtr nativeCallback, Status status)
+        {
+            Logger.LogDefault(LogLevel.Trace, $"SyncSocketProvider running native callback {nativeCallback} with status {status.Code} \"{status.Reason}\".");
+
+            if (!string.IsNullOrEmpty(status.Reason))
+            {
+                var bytes = Encoding.UTF8.GetBytes(status.Reason);
+                unsafe
+                {
+                    fixed (byte* data = bytes)
+                    {
+                        var reason = new StringValue { data = data, size = bytes.Length };
+                        NativeMethods.run_callback(nativeCallback, status.Code, reason);
+                    }
+                }
+            }
+            else
+            {
+                NativeMethods.run_callback(nativeCallback, status.Code, StringValue.Null);
+            }
+        }
+
+        private async Task PostWorkAsync(IntPtr nativeCallback)
+        {
+            Logger.LogDefault(LogLevel.Trace, "Posting work to SyncSocketProvider event loop.");
+            await _workQueue.Writer.WriteAsync(new EventLoopWork(nativeCallback, Status.OK, _cts.Token));
+        }
+
+        private partial async Task WorkThread()
+        {
+            Logger.LogDefault(LogLevel.Trace, "Starting SyncSocketProvider event loop.");
+            try
+            {
+                while (await _workQueue.Reader.WaitToReadAsync())
+                {
+                    while (_workQueue.Reader.TryRead(out var work))
+                    {
+                        work.Execute();
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.LogDefault(LogLevel.Error, $"Error occurred in SyncSocketProvider event loop {e.GetType().FullName}: {e.Message}");
+                Logger.LogDefault(LogLevel.Trace, e.StackTrace);
```

Comment:
There isn't an interface we can use to notify Sync about that. But now that you mention it the existing behavior on an exception on the sync worker thread is to bring down the process, so instead of swallowing and logging the exception that's what we should do here.